DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1CompareTrackSoA.cc

```diff
@@ -0,0 +1,200 @@
+  MonitorElement* hnLayersVsPhi;
+  MonitorElement* hnLayersVsEta;
+  MonitorElement* hchi2;
+  MonitorElement* hChi2VsPhi;
+  MonitorElement* hChi2VsEta;
+  MonitorElement* hpt;
+  MonitorElement* heta;
+  MonitorElement* hphi;
+  MonitorElement* hz;
+  MonitorElement* htip;
+  MonitorElement* hquality;
+};
+
+//
+// constructors
+//
+
+SiPixelPhase1CompareTrackSoA::SiPixelPhase1CompareTrackSoA(const edm::ParameterSet& iConfig)
+    : tokenSoATrackCPU_(consumes<PixelTrackHeterogeneous>(iConfig.getParameter<edm::InputTag>("pixelTrackSrcCPU"))),
+      tokenSoATrackGPU_(consumes<PixelTrackHeterogeneous>(iConfig.getParameter<edm::InputTag>("pixelTrackSrcGPU"))),
+      topFolderName_(iConfig.getParameter<std::string>("TopFolderName")),
+      useQualityCut_(iConfig.getParameter<bool>("useQualityCut")),
+      minQuality_(pixelTrack::qualityByName(iConfig.getParameter<std::string>("minQuality"))),
+      dr2cut_(iConfig.getParameter<double>("deltaR2cut")) {}
+
+//
+// -- Analyze
+//
+void SiPixelPhase1CompareTrackSoA::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  const auto& tsoaHandleCPU = iEvent.getHandle(tokenSoATrackCPU_);
+  const auto& tsoaHandleGPU = iEvent.getHandle(tokenSoATrackGPU_);
+  if (!tsoaHandleCPU.isValid() || !tsoaHandleGPU) {
+    edm::LogWarning("SiPixelPhase1CompareTrackSoA")
+        << "Either GPU or CPU tracks not found! Hence comparison not run!" << std::endl;
+    return;
+  }
+
+  auto const& tsoaCPU = *((tsoaHandleCPU.product())->get());
+  auto const& tsoaGPU = *((tsoaHandleGPU.product())->get());
+  auto maxTracksCPU = tsoaCPU.stride();  //this should be same for both?
+  auto maxTracksGPU = tsoaGPU.stride();  //this should be same for both?
+  auto const* qualityCPU = tsoaCPU.qualityData();
+  auto const* qualityGPU = tsoaGPU.qualityData();
+  int32_t nTracksCPU = 0;
+  int32_t nTracksGPU = 0;
+  int32_t nLooseAndAboveTracksCPU = 0;
+  int32_t nLooseAndAboveTracksCPU_matchedGPU = 0;
+  int32_t nLooseAndAboveTracksGPU = 0;
+
+  //Loop over GPU tracks and store the indices of the loose tracks. Whats happens if useQualityCut_ is false?
+  std::vector<int32_t> looseTrkidxGPU;
+  for (int32_t jt = 0; jt < maxTracksGPU; ++jt) {
+    if (tsoaGPU.nHits(jt) == 0)
+      break;  // this is a guard
+    if (!(tsoaGPU.pt(jt) > 0.))
+      continue;
+    nTracksGPU++;
+    if (useQualityCut_ && qualityGPU[jt] < minQuality_)
+      continue;
+    nLooseAndAboveTracksGPU++;
+    looseTrkidxGPU.emplace_back(jt);
+  }
+
+  //Now loop over CPU tracks//nested loop for loose gPU tracks
+  for (int32_t it = 0; it < maxTracksCPU; ++it) {
+    if (tsoaCPU.nHits(it) == 0)
+      break;  // this is a guard
+    if (!(tsoaCPU.pt(it) > 0.))
+      continue;
+    nTracksCPU++;
+    if (useQualityCut_ && qualityCPU[it] < minQuality_)
+      continue;
+    nLooseAndAboveTracksCPU++;
+    //Now loop over loose GPU trk and find the closest in DeltaR//do we need pt cut?
+    int32_t closestTkidx = 99999;
```

Comment:
I would suggest defining a constant for "invalid" or "not found" and using it consistently:
```suggestion
    const int32_t notFound = -1;
    int32_t closestTkidx = notFound;
    
    ...
    
    if (closestTkidx == notFound)
      continue;
```

Reply:
done at [76b4e67](https://github.com/cms-sw/cmssw/pull/37619/commits/76b4e679e9fa98f05f5fc7095d00eb4b03926ef5)